src/source/Ice/SocketConnection.c

```diff
@@ -96,54 +82,67 @@ STATUS freeSocketConnection(PSocketConnection* ppSocketConnection)
 
-    CHK(pSocketConnection != NULL, STATUS_NULL_ARG);
-
-    pSocketConnection->pSslCtx = SSL_CTX_new(SSLv23_method());
-
-    CHK(pSocketConnection->pSslCtx != NULL, STATUS_SSL_CTX_CREATION_FAILED);
-
-    SSL_CTX_set_read_ahead(pSocketConnection->pSslCtx, 1);
-    SSL_CTX_set_verify(pSocketConnection->pSslCtx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, certificateVerifyCallback);
+    pSocketConnection = (PSocketConnection) customData;
+    retStatus = socketSendDataWithRetry(pSocketConnection, pBuffer, bufferLen, NULL, NULL);
 
-    CHK(SSL_CTX_set_cipher_list(pSocketConnection->pSslCtx, "HIGH:!aNULL:!MD5:!RC4"), STATUS_SSL_CTX_CREATION_FAILED);
+CleanUp:
+    return retStatus;
+}
 
-    pSocketConnection->pSsl = SSL_new(pSocketConnection->pSslCtx);
-    CHK(pSocketConnection->pSsl != NULL, STATUS_CREATE_SSL_FAILED);
+static VOID tlsSessionOnStateChange(UINT64 customData, TLS_SESSION_STATE state)
+{
+    PSocketConnection pSocketConnection = NULL;
+    if (customData == 0) {
+        return;
+    }
 
-    if (isServer) {
-        SSL_set_accept_state(pSocketConnection->pSsl);
-    } else {
-        SSL_set_connect_state(pSocketConnection->pSsl);
+    pSocketConnection = (PSocketConnection) customData;
+    switch (state) {
+        case TLS_SESSION_STATE_NEW:
+            pSocketConnection->tlsHandshakeStartTime = INVALID_TIMESTAMP_VALUE;
+            break;
+        case TLS_SESSION_STATE_CONNECTING:
+            pSocketConnection->tlsHandshakeStartTime = GETTIME();
+            break;
+        case TLS_SESSION_STATE_CONNECTED:
+            if (IS_VALID_TIMESTAMP(pSocketConnection->tlsHandshakeStartTime)) {
+                DLOGD("TLS handshake done. Time taken %" PRIu64 " ms",
+                      (GETTIME() - pSocketConnection->tlsHandshakeStartTime) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+                pSocketConnection->tlsHandshakeStartTime = INVALID_TIMESTAMP_VALUE;
+            }
+            break;
+        case TLS_SESSION_STATE_CLOSED:
+            ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+            break;
     }
+}
 
-    SSL_set_mode(pSocketConnection->pSsl, SSL_MODE_AUTO_RETRY);
-    CHK((pSocketConnection->pReadBio = BIO_new(BIO_s_mem())) != NULL, STATUS_SSL_CTX_CREATION_FAILED);
-    CHK((pSocketConnection->pWriteBio = BIO_new(BIO_s_mem())) != NULL, STATUS_SSL_CTX_CREATION_FAILED);
+STATUS socketConnectionInitSecureConnection(PSocketConnection pSocketConnection, BOOL isServer)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    
+    CHK(pSocketConnection != NULL, STATUS_NULL_ARG);
+    CHK(pSocketConnection->pTlsSession == NULL, STATUS_INVALID_ARG);
 
-    BIO_set_mem_eof_return(pSocketConnection->pReadBio, -1);
-    BIO_set_mem_eof_return(pSocketConnection->pWriteBio, -1);
-    SSL_set_bio(pSocketConnection->pSsl, pSocketConnection->pReadBio, pSocketConnection->pWriteBio);
-    pSocketConnection->freeBios = FALSE;
+    TlsSessionCallbacks callbacks;
```

Comment:
Move all local stack declarations before any code that can "jump" over it. This will generate an error on pure c compilers

Reply:
My bad, I forgot to do that. I'll update it. Thanks.